src/tasks/withdraw.ts

```diff
@@ -246,36 +247,38 @@ async function getWithdrawals(
   leftover: string,
   hre: HardhatRuntimeEnvironment,
 ): Promise<Withdrawal[]> {
-  const withdrawals = [];
+  const withdrawals: Withdrawal[] = [];
   const minValueWei = utils.parseUnits(minValue, DAI_DECIMALS);
   const leftoverWei = utils.parseUnits(leftover, DAI_DECIMALS);
-  for (let i = 0; i < tokens.length; i++) {
+  let vanishingProgressMessage = "";
+  const addWithdrawal = async (i: number) => {
     const address = tokens[i];
-    clearLine();
-    process.stdout.write(`Processing token ${i + 1}/${tokens.length}...`);
     const token = await fastTokenDetails(address, hre);
     const balance = await token.contract.balanceOf(settlement.address);
     if (balance.eq(0)) {
-      continue;
+      return;
     }
     const pricedToken = await appraise(token, hre.network.name);
     const balanceUsd = pricedToken.usdValue
       .mul(balance)
       .div(BigNumber.from(10).pow(pricedToken.decimals));
-    clearLine();
     // Note: if balanceUsd is zero, then setting either minValue or leftoverWei
     // to a nonzero value means that nothing should be withdrawn. If neither
     // flag is set, then whether to withdraw does not depend on the USD value.
     if (
       balanceUsd.lt(minValueWei.add(leftoverWei)) ||
       (balanceUsd.isZero() && !(minValueWei.isZero() && leftoverWei.isZero()))
     ) {
+      clearLine();
       console.warn(
         `Ignored ${utils.formatUnits(balance, pricedToken.decimals)} units of ${
           token.symbol ?? "unknown token"
         } (${token.address}) with value ${formatUsdValue(balanceUsd)} USD`,
       );
-      continue;
+      if (process.stdin.isTTY) {
+        process.stdout.write(vanishingProgressMessage);
+      }
+      return;
```

Comment:
Is this the progress bar? I don't know what TTY is and my google search didn't help much ("tty stands for TeleTYpewriter.") but I will come back here in the future when I need a TS progress bar.

Reply:
There are definitely subtleties in this that I don't know of, but here is my mental model about what's going on:
TTY is linux lingo for "interactive terminal". Every time you are running a command in the console (e.g., `$ node`), it's attached to the standand input and output (and more) of the console, which is an interactive terminal and so `process.stdin.isTTY == true`. The standard output can also be a file if you redirect it (`>`) or pipe it (you can see the difference with `$ node | tee /dev/null`), in this case it follows different rules, for example some things are not available like being able to move the cursor or clear the line. That's why for example the output of `$ ls` and `$ ls | cat` is not the same (colors missing, newlines instead of column formatting).
So what's going on here is that I write a line to stdout directly, and before printing an updated one I erase it (`clearLine`). That's a bit hacky by the way, because if I use `console.log` somewhere else in the code then the log is written on the same line and the progress text cannot be cleared as easily anymore.